Automattic/newspack-newsletters
# feat: active campaign segments support

### All Submissions:






### Changes proposed in this Pull Request:

Implements support for sending newsletters to Active Campaign's segments.

Active Campaign currently supports APIs v1 and v3 and to implement this feature we must use both versions. v1 does not have a method for fetching existing segments whereas v3 does not have a method to create campaigns. To handle that in an organized manner, `api_v1_request()` and `api_v3_request()` methods were implemented.

This ESP has campaigns and messages, a campaign can hold multiple messages and choose which message to send on the scheduling/sending action of the campaign. The current sync action creates a message and a campaign, attaching the message to the campaign. This is problematic because neither API (v1 or v3) has an endpoint for editing a campaign. Once a sync occurs with a specific target (sender and list) it cannot be edited through us.

This PR also changes this behavior to only sync the message while editing, the campaign is only created on sending. This fixes the issue by not having to edit a campaign prior to sending.



Closes #812.

### How to test the changes in this Pull Request:

1. Connect to an existing Active Campaign account that has segments
2. Create a new newsletter and select a list and a segment to send to
3. Send and confirm on the ESP dashboard that it was sent to the selected list and segment

### Other information:
